nixos/gnome-keyring: rewrite module and fix unlocking on GDM session login

JohnRTitor opened the pull request:
## Description of changes

This PR rewrites the `nixos/gnome/gnome-keyring` module.

GDM uses gdm-password as the PAM service name for both logins and unlocks.
So unlock gnome-keyring as part of `gdm-password`.
Without this, keyrings may not be unlocked properly for GDM 45+.

## Things done

- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

@NixOS/gnome

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

JohnRTitor committed:
nixos/gnome/gnome-keyring: rewrite the implementation

JohnRTitor commented:
Building NixOS manual check [fails](https://github.com/NixOS/nixpkgs/actions/runs/9302524845/job/25602901205?pr=315874), any idea why?

jtojnar commented:
Docs do not allow `pkgs` references since they would prevent the docs from being cacheable. You need to include `defaultText` for the manual when you use a restricted value in `default`.

jtojnar commented on nixos/modules/services/desktops/gnome/gnome-keyring.nix:
```diff
@@ -1,53 +1,53 @@
-          such as user names and passwords.
-        '';
-      };
+  ## Interface ##
```
We are removing these comments in GNOME modules when refactoring, since they are redundant with `options` and `config` attributes. We still use `options = {}` and `meta` unpacked to make the sections more visible.

jtojnar commented on nixos/modules/services/desktops/gnome/gnome-keyring.nix:
```diff
@@ -1,53 +1,52 @@
 
+  options.services.gnome.gnome-keyring = {
+    enable = lib.mkEnableOption null // {
+      default = false;
```
This is the default `default` for `mkEnableOption`.

Why not just use

```nix
lib.mkEnableOption ''
        GNOME Keyring daemon, a service designed to
        take care of the user's security credentials,
        such as user names and passwords
''

jtojnar commented on nixos/modules/services/desktops/gnome/gnome-keyring.nix:
```diff
@@ -1,53 +1,52 @@
+      '';
     };
-
+    package = lib.mkPackageOption pkgs "gnome-keyring" { default = pkgs.gnome.gnome-keyring; };
```
Is there as specific reason for adding the package option? Are users expected to change it? If it is just for debugging, maybe overlays would be sufficient?

JohnRTitor replied:
Overlays are pretty complicated for new users. I added the option in case anyone wants to override the gnome-keyring package like use an older version, or patch it themselves.

This package option helps sort that out without resorting to overlays, which can actually rebuild the whole system sometimes (may be not for this package). Overlays are known to cause problems.

As for the `gcrPackage` option, I thought I would use `gcr_4` with this. I assume `gcr` ver 3 is still the default due to compatibility issues but for GNOME 46, shouldn't `gcr_4` be the default?

jtojnar replied:
Well, new users should not need to change packages under modules in the first place. If there is a bug, we should fix it in Nixpgks. And the mass rebuilds are not an issue with leaf packages like GNOME Keyring.

`gcr` contains two D-Bus services that were removed in [GCR 4](https://gitlab.gnome.org/GNOME/gcr/-/commit/f316c8919bfcbe2534271e9729084342db3cc9c6). AFAICT, it does not even make sense to install that on GNOME Shell, since it provides its own implementation.

Looks like `gcr` installation was introduced in 2d280840f844cad6c9478bc9107b02cff82b1c34 to provide fallback prompter for GNOME Keyring (well, it uses the prompter through `gcr` library). But that should probably go into its own module, as GNOME Keyring is [not the only software that expects this](https://codesearch.debian.net/search?q=org.gnome.keyring.SystemPrompter&literal=1).

JohnRTitor replied:
Done. Removed `package` and `gcrPackage` option.

jtojnar commented:
> GDM uses gdm-password as the PAM service name for both logins and unlocks.
> So unlock gnome-keyring as part of `gdm-password`.
> Without this, keyrings may not be unlocked properly for GDM 45+.

Do you have some tips for reproducing this? Keyring unlocks fine for me on 46.

---

The whole GDM PAM setup would deserve a re-do but PAM feels like black magic, and the man page probably contains a curse that erases my memory every time I look at it.

JohnRTitor commented:
> Do you have some tips for reproducing this? Keyring unlocks fine for me on 46.

I am not actually using GNOME 46. My setup is Hyprland window manager with a set of GNOME packages.
Keyrings are not automatically unlocked for me and I have to do that manually unless I use this option.

Other users on NixOS discourse have noticed this issue as well. See https://discourse.nixos.org/t/login-keyring-did-not-get-unlocked-hyprland/40869/11

jtojnar commented on nixos/modules/services/desktops/gnome/gnome-keyring.nix:
```diff
@@ -1,53 +1,48 @@
+      enable = lib.mkEnableOption ''
+        GNOME Keyring daemon, a service designed to
+        take care of the user's security credentials,
+        such as user names and passwords.
```
mkEnableOption already includes period

jtojnar commented on nixos/modules/services/desktops/gnome/gnome-keyring.nix:
```diff
@@ -1,53 +1,48 @@
-  config = lib.mkIf config.services.gnome.gnome-keyring.enable {
-
-    environment.systemPackages = [ pkgs.gnome.gnome-keyring ];
+    services.dbus.packages = with pkgs; [
```
`with` makes reference checking lazy so it can hide errors. I would avoid it here since it does not add much benefit here and actually makes the code slightly longer.

jtojnar commented on nixos/modules/services/desktops/gnome/gnome-keyring.nix:
```diff
@@ -36,6 +36,7 @@ in
     xdg.portal.extraPortals = with pkgs; [ gnome.gnome-keyring ];
 
     security.pam.services.login.enableGnomeKeyring = true;
+    security.pam.services.gdm-password.enableGnomeKeyring = config.services.xserver.displayManager.gdm.enable;
```
Arch adds this as well so it is probably fine https://gitlab.gnome.org/GNOME/gdm/-/blob/main/data/pam-arch/gdm-password.pam#L4

JohnRTitor replied:
Nice. But I believe this should only be enabled if GDM is used?

Some users (without GNOME) may prefer another login manager such as SDDM, in which case `login` PAM should work.

JohnRTitor replied:
I enabled gnome-keyring for `autologin`, `fingerprint` and `smartcard` authentication mode too.

jtojnar replied:
Right, I just meant that it should be okay to include keyring in `gdm-password`. Will need to double check the other auth modes, in case their presence signals specific behavour to GDM (we previously did not have those files).

JohnRTitor replied:
I can't test fingerprint and smartcard, but autologin seems fine.

JohnRTitor commented:
@jtojnar Would love to see this patch in 24.05 too since it's also broken there. 

Also, have you checked other authentication modes such as fingerprint and smartcard? 

I don't think these should cause any problems. Arch seems to run this just fine.

https://gitlab.gnome.org/GNOME/gdm/-/blob/main/data/pam-arch/gdm-fingerprint.pam#L10
https://gitlab.gnome.org/GNOME/gdm/-/blob/main/data/pam-arch/gdm-smartcard.pam#L10

jtojnar commented:
> Also, have you checked other authentication modes such as fingerprint and smartcard?
> 
> I don't think these should cause any problems. Arch seems to run this just fine.

I meant that we do not have those files currently. And adding `security.pam.services.gdm-fingerprint.enableGnomeKeyring = true;` creates the file with only some minimal rules.

I am not confident enough with PAM to be able to predict the effects it will have.

Wanted to look deeper into PAM this weekend but stuff got in the way. Finding someone who understands PAM and/or GDM for review would make this faster.

> Would love to see this patch in 24.05 too since it's also broken there.

Probably would want to run it in unstable for a bit to catch possible issues.

JohnRTitor commented:
> Wanted to look deeper into PAM this weekend but stuff got in the way. Finding someone who understands PAM and/or GDM for review would make this faster.

Sure but I am not sure who to tag, but probably @nbraud and @LeSuisse who were involved with the last PR to `nixos/pam` (#277626),

> Probably would want to run it in unstable for a bit to catch possible issues.

Sure.

SuperSandro2000 commented on nixos/modules/services/desktops/gnome/gnome-keyring.nix:
```diff
@@ -1,53 +1,54 @@
-    security.pam.services.login.enableGnomeKeyring = true;
+    security.pam.services = {
+      login.enableGnomeKeyring = true;
+    } // lib.optionalAttrs config.services.xserver.displayManager.gdm.enable {
```
Shouldn't we use mkMerge here to leverage the module system?

JohnRTitor replied:
Is there any benefits in that? I did this way as this looks cleaner.

JohnRTitor replied:
@SuperSandro2000 Done. Please check.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4056

JohnRTitor commented:
I feel like getting this to be reviewed and tested will take a lot of time. @jtojnar so can we just drop all changes to PAM except for gdm-password, and get this merged?

LeSuisse commented on nixos/modules/services/desktops/gnome/gnome-keyring.nix:
```diff
@@ -1,53 +1,57 @@
+        gdm-password.enableGnomeKeyring = true;
+        gdm-autologin.enableGnomeKeyring = true;
+        gdm-fingerprint.enableGnomeKeyring = true;
+        gd-smartcard.enableGnomeKeyring = true;
```
```suggestion
        gdm-smartcard.enableGnomeKeyring = true;
```

Also, it might reasonable to not set it right now, there is some discussions about `gdm-smartcard` and unwanted interaction with Yubikeys (seems to concern at least some Arch and Debian users).


https://gitlab.gnome.org/GNOME/gdm/-/issues/877

JohnRTitor replied:
Sure. I have removed the smartcard bit. Can be enabled later if proven stable.

LeSuisse commented on nixos/modules/services/desktops/gnome/gnome-keyring.nix:
```diff
@@ -1,53 +1,57 @@
+      (lib.mkIf config.services.xserver.displayManager.gdm.enable {
+        gdm-password.enableGnomeKeyring = true;
+        gdm-autologin.enableGnomeKeyring = true;
+        gdm-fingerprint.enableGnomeKeyring = true;
```
It might be reasonable to only enable this automatically when `config.services.fprintd.enable` is enabled. It is unlikely to be useful in other scenario (or in the worst case the users can still enable it themselves like today).

JohnRTitor replied:
Done, anything else I should check for?

JohnRTitor committed:
nixos/gnome-keyring: unlock keyring with gdm-password, gdm-autologin

GDM uses gdm-password as the PAM service name for both logins and unlocks.
So unlock gnome-keyring as part of `gdm-password`.
Without this, keyrings may not be unlocked properly for GDM 45+.

also unlock as part of GDM autologin

JohnRTitor committed:
nixos/gnome-keyring: enable gnome-keyring for fingerprint authentication'

this should be enabled by default if fprintd is enabled

LeSuisse commented on nixos/modules/services/desktops/gnome/gnome-keyring.nix:
```diff
@@ -1,53 +1,56 @@
+      {
+        login.enableGnomeKeyring = true;
+      }
+      (lib.mkIf config.services.xserver.displayManager.gdm.enable {
+        gdm-password.enableGnomeKeyring = true;
+        gdm-autologin.enableGnomeKeyring = true;
+        gdm-fingerprint.enableGnomeKeyring = config.services.fprintd.enable;
+      })
```
```suggestion
      (lib.mkIf config.services.xserver.displayManager.gdm.enable {
        gdm-password.enableGnomeKeyring = true;
        gdm-autologin.enableGnomeKeyring = true;
      })
      (lib.mkIf (config.services.xserver.displayManager.gdm.enable && config.services.fprintd.enable) {
        gdm-fingerprint.enableGnomeKeyring = true;
      })
```

That will avoid the creation of a `gdm-fingerprint` pam config with a default deny if we do not need it.

JohnRTitor replied:
Done.

JohnRTitor commented:
@LeSuisse can we merge this?

As for fingerprint authentication, I plan to cherry-pick https://github.com/NixOS/nixpkgs/pull/306338 in a different PR

LeSuisse reviewed (approved):
Yum the pam config looks reasonable to me at this stage
